### PR TITLE
Fluids - Minor Refactor

### PR DESCRIPTION
**Pull request by LeilaGhaffari**
This PR aims at removing `computeFluxJacobian_NS(CeedScalar dF[3][5][5], ...)` and instead, using variational notation for computing the Flux Jacobian.

### CODE REVIEW

**Comment by jedbrown** [comment]
I haven't forgotten about this (which I think is a valuable refactor). Can you just ping when you're ready to revisit this, I assume after #1011 merges and this is rebased (or merge from main into this branch)?

**Comment by LeilaGhaffari** [comment]
> I haven't forgotten about this (which I think is a valuable refactor). Can you just ping when you're ready to revisit this, I assume after #1011 merges and this is rebased (or merge from main into this branch)?

Sure, I will update/fix the remaining parts (and rebase it onto main) after #1011 merges.

**Comment by LeilaGhaffari** [comment]
Alright, this is ready for review.

File: examples/fluids/qfunctions/channel.h
Comment (jrwrigh) [inline examples/fluids/qfunctions/channel.h:122]: Wouldn't this also set `g` to zero for the other functions too? Making the volumetric forcing term not work for periodic BCs?

File: examples/fluids/qfunctions/channel.h
Comment (jrwrigh) [inline examples/fluids/qfunctions/channel.h:90]: I like the `UnpackState*` functions! I think we could move this to a function pointer method similar to what I did with the `newtonian.h` boundary integral QFs, where we set a generic `SetStateFromQi` function at the start of the routine and just use that throughout, `qi` representing whatever state variables the solver is using at the time.

File: examples/fluids/qfunctions/channel.h
Comment (jrwrigh) [reply examples/fluids/qfunctions/channel.h:122]: If the attempt is to remove the potential energy, I think you could just set `x[0] = 0`.

File: examples/fluids/qfunctions/newtonian.h
Comment (jrwrigh) [inline examples/fluids/qfunctions/newtonian.h:389]: Any strong preference for curly braces vs not? I kinda prefer without when possible. There are a few other places in the code where we don't have curly braces on double for-loops, which is mainly why I ask.

File: examples/fluids/qfunctions/stabilization.h
Comment (jrwrigh) [inline examples/fluids/qfunctions/stabilization.h:57]: Can probably just be `Tau_diag` since it's using a `State` input now.

File: examples/fluids/qfunctions/newtonian.h
Comment (jrwrigh) [inline examples/fluids/qfunctions/newtonian.h:732]: I know I had asked @jedbrown about removing this before (see [here](https://github.com/CEED/libCEED/pull/962#discussion_r889586560)). Not sure if the same reasoning applies now though. Jed mentions refactoring this block of code, so maybe this is good to go now.

File: examples/fluids/qfunctions/stabilization.h
Comment (jrwrigh) [inline examples/fluids/qfunctions/stabilization.h:14]:
```suggestion
#include "newtonian_state.h"
#include <ceed.h>
```

Stop some unnecessary warnings from my LSP.

File: examples/fluids/qfunctions/stabilization.h
Comment (jrwrigh) [inline examples/fluids/qfunctions/stabilization.h:131]: Do we need to keep this?

File: examples/fluids/qfunctions/stabilization.h
Comment (jrwrigh) [inline examples/fluids/qfunctions/stabilization.h:134]:
```suggestion
  CeedScalar u_sq = Dot3(u,u); // Velocity square
```

File: examples/fluids/qfunctions/densitycurrent.h
Comment (jrwrigh) [inline examples/fluids/qfunctions/densitycurrent.h:159]: Could do similar function pointer stuff.

**Review by jrwrigh — APPROVED** [review]
Few suggestions. Otherwise, looks good to me, and a few extra comments:

1) There are `__attribute ((unused))` by the initializations of `Tau_d` (https://github.com/CEED/libCEED/blob/779062404f6c183e987cbc971fc4cdeea0db0363/examples/fluids/qfunctions/newtonian.h#L431 and https://github.com/CEED/libCEED/blob/779062404f6c183e987cbc971fc4cdeea0db0363/examples/fluids/qfunctions/newtonian.h#L1063). I don't think these are necessary anymore?

2) Might be worth switching to Function pointers for `State` related functions for `IFunction_Newtonian`. Then we can move to only having one IFunction instead of 2. The only difference between them now is different treatments of `q_dot`. I don't think that extra bit of efficiency is worth keeping them separate. But that can also be a different PR, as I'm not sure if how I currently implemented the function pointers is good or not (or really if it can be improved).

File: examples/fluids/qfunctions/channel.h
Comment (jedbrown) [reply examples/fluids/qfunctions/channel.h:90]: I'd want to do a performance test before using function pointers in perf-critical code (like the main volume integrals). I'm less concerned elsewhere. It's possible to transform so the key function pointers are passed as arguments to a function that is meant to be inlined (topic of another PR if we want to go that route).

File: examples/fluids/qfunctions/newtonian.h
Comment (jedbrown) [reply examples/fluids/qfunctions/newtonian.h:389]: Not a big deal to me. Rust requires them always.

Comment (jeremylt) [reply examples/fluids/qfunctions/newtonian.h:389]: In Ratel I adopted the PETSc convention - {} required for all multi line statements.

So
good:
```C
for (i = 0; i < 42; i++) total += array[i];
```
good:
```C
for (i = 0; i < 42; i++) {
  total += array[i];
}
```
bad:
```C
for (i = 0; i < 42; i++)
  total += array[i];
```

I mostly did that because I have screwed it up a couple of times

File: examples/fluids/qfunctions/newtonian.h
Comment (jedbrown) [inline examples/fluids/qfunctions/newtonian.h:485]: I guess one way to write this would be to have NONE/SU/SUPG just change what terms are active on this line, and then you could remove the switch down below. The current structure is fine too.

**Review by jedbrown — APPROVED** [review]

File: examples/fluids/qfunctions/stabilization.h
Comment (jedbrown) [reply examples/fluids/qfunctions/stabilization.h:131]: It doesn't get called. We might want such a thing when implementing an eigenvalue/matrix-sqrt stabilization operator, though we can resurrect it from the history if we don't have a better way to compute the matrix at that time.

File: examples/fluids/qfunctions/stabilization.h
Comment (jedbrown) [reply examples/fluids/qfunctions/stabilization.h:57]: I think the point is that physically, it's meant to take a (conservative) residual and produce fluctuations Y' (primitive). The `Prim` isn't about how the input state is represented, but that the output (projected subgrid fluctuations) Y' is primitive.

Comment (jedbrown) [reply examples/fluids/qfunctions/stabilization.h:57]: The `NOT UPDATED YET` can presumably be removed.

File: examples/fluids/qfunctions/newtonian.h
Comment (jedbrown) [reply examples/fluids/qfunctions/newtonian.h:732]: :shrug: Not important to me any more, as our stabilization code has been exercised more.

File: examples/fluids/qfunctions/stabilization.h
Comment (jrwrigh) [inline examples/fluids/qfunctions/stabilization.h:44]: Should we move the zeroing of `stab[][]` into the function itself. Seems like a simple thing that might prevent a headache in the future.

File: examples/fluids/qfunctions/densitycurrent.h
Comment (jrwrigh) [reply examples/fluids/qfunctions/densitycurrent.h:159]: See https://github.com/CEED/libCEED/pull/995#discussion_r927745566

File: examples/fluids/qfunctions/channel.h
Comment (LeilaGhaffari) [reply examples/fluids/qfunctions/channel.h]: I just confirmed that it doesn't affect other QFunctions but the reason I chose to set g to zero is that I thought it was more intuitive and could help with the readability of the code. I was also concerned about  nonzero g[2] and g[2] but I am seeing now that they are set to zero. Anyways, just changed it in [this commit](https://github.com/CEED/libCEED/pull/995/commits/f712d7f79c535eaa1b03e0b87d72a30fbb5c7840) with more comments.

File: examples/fluids/qfunctions/newtonian.h
Comment (LeilaGhaffari) [reply examples/fluids/qfunctions/newtonian.h:389]: I was trying to make the style consistent throughout that file but if you are seeing more of those curly braces, then I have missed them.

Comment (jeremylt) [reply examples/fluids/qfunctions/newtonian.h:389]: haha, I have been trying to push code I write more towards the style *with* braces instead of *without* braces.

File: examples/fluids/qfunctions/stabilization.h
Comment (LeilaGhaffari) [reply examples/fluids/qfunctions/stabilization.h:44]: Good call!

File: examples/fluids/qfunctions/newtonian.h
Comment (LeilaGhaffari) [reply examples/fluids/qfunctions/newtonian.h:389]: I agree with you, @jeremylt , that the last style is more error-prone and I am happy to stick to the first two if we want to add it to our style convention.

Comment (jedbrown) [reply examples/fluids/qfunctions/newtonian.h:389]: I think clang-format can do this automatically if we migrate there at some point.

File: examples/fluids/qfunctions/newtonian.h
Comment (LeilaGhaffari) [reply examples/fluids/qfunctions/newtonian.h]: What do you think of [the last commit](https://github.com/CEED/libCEED/pull/995/commits/b70ae9d3e37855b20b547f7fe4ba07cd52b70e4b)?

File: examples/fluids/qfunctions/stabilization.h
Comment (jedbrown) [inline examples/fluids/qfunctions/stabilization.h:72]: To avoid confusion, the `Y` here has units of `[conservative]/[time]` so it should have a different name.

File: examples/fluids/qfunctions/newtonian.h
Comment (jedbrown) [reply examples/fluids/qfunctions/newtonian.h]: Great, modulo one comment I left there.

File: examples/fluids/qfunctions/stabilization.h
Comment (LeilaGhaffari) [reply examples/fluids/qfunctions/stabilization.h:72]: Oh, right. I meant to change that. `U_dot`?

Comment (LeilaGhaffari) [reply examples/fluids/qfunctions/stabilization.h:72]: Nope, we have that already!

File: examples/fluids/qfunctions/channel.h
Comment (jrwrigh) [reply examples/fluids/qfunctions/channel.h]: > I just confirmed that it doesn't affect other QFunctions

Interesting. Does libCEED make a copy of the struct if you don't set `CeedQFunctionSetContextWritable`? @jedbrown @jeremylt

File: examples/fluids/qfunctions/stabilization.h
Comment (jedbrown) [reply examples/fluids/qfunctions/stabilization.h:72]: `R` for residual is common in the literature.

File: examples/fluids/qfunctions/channel.h
Comment (jeremylt) [reply examples/fluids/qfunctions/channel.h]: It shouldn't. If you don't set the context read-only, then changes should persist. Could be that this ctx has a copy of the struct in creation

Comment (jedbrown) [reply examples/fluids/qfunctions/channel.h]: I'd thought this was copied out somewhere in the qfunctions, but perhaps not. I don't immediately know why this wouldn't affect other qfunctions.

File: examples/fluids/qfunctions/newtonian.h
Comment (LeilaGhaffari) [reply examples/fluids/qfunctions/newtonian.h]: Cool, just fixed renamed it.

**Review by jedbrown — APPROVED** [review]
Looks good to me now.

**Comment by LeilaGhaffari** [comment]
Thanks! I am happy to `squash-merge` if there are no objections.

**Comment by jedbrown** [comment]
No objections from me as soon as the last test passes.

**Comment by LeilaGhaffari** [comment]
Sorry, I had messed up the sign for stabilization terms in `RHSFunction_Newtonian()` (caused by copy/pasting) but I think it is ready to merge after the tests pass.

**Pull request merged**
